js/formats/types.ts

```diff
@@ -212,9 +212,16 @@ export interface DiskProcessedResponse {
 export type FormatWorkerResponse =
     DiskProcessedResponse;
 
+export interface MassStorageData {
+    name: string;
+    ext: string;
+    data: ArrayBuffer;
```

Comment:
I'm confused about why this is typed as `ArrayBuffer` instead of `Uint8Array`? Up in `smartport.ts`, you have:

```typescript
    getBinary(drive: number): MassStorageData {
        const blocks = this.disks[drive].blocks;
        const data = new Uint8Array(blocks.length * 512);
        for (let idx = 0; idx < blocks.length; idx++) {
            data.set(blocks[idx], idx * 512);
        }
        return {
            name: this.disks[drive].name,
            ext: this.ext[drive],
            data,
        };
    }
```

So it looks like `data` should be a `Uint8Array`. Is there a reason not to?

Reply:
That it's a `Uint8Array` is an implementation detail, you can't construct an `ArrayBuffer` directly. `ArrayBuffer` is the interface that's being used when passing around raw bytes that will eventually be consumed with a `DataView`. This makes it symmetric with the `setBinary()` interface.

Reply:
But `Uint8Array` isn't even an `ArrayBuffer`—you have to get it from its `buffer` property:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array#instance_properties

It just happens that `Uint8Array` has the same structure as `ArrayBuffer`, so TypeScript lets you assign a `Uint8Array` to a variable declared `ArrayBuffer`. And most of the time that works because you can pass a `Uint8Array` to a `Uint8Array` constructor. But if pass it to a `DataView` constructor:

```javascript
a = new Uint8Array(100);
d = new DataView(a);
// Uncaught TypeError: First argument to DataView constructor must be an ArrayBuffer
```

And this is what would happen if you tried to `new DataView(data)` from the code I quoted above.

So, I agree that you could always use `ArrayBuffer`, but then you're constantly creating new `Uint8Array` wrappers (as that seems the most common use case). Otherwise you could pass around `Uint8Array` and use `new DataView(a.buffer)` when you need to.

Reply:
It sort of feels like a wash, the interface to various APIs are all all `ArrayBuffers`, all the non-bitstream activity is using `DataView`s, and byte manipulation is done with `Uint8Arrays`. Since it's just a view and not copy, creating new arrays is cheap.

It is particularly annoying that Typescript can't tell them apart, though.

Reply:
Yes, indeed. Apparently this has been outstanding a long time: https://github.com/Microsoft/Typescript/issues/202 I don't know that there's a good solution for the moment that's not rather invasive.

If you want to stick with `ArrayBuffer`, I'm fine with that.